File: go/circuit_breaker.go
package devlogs

import (
	"fmt"
	"os"
	"sync"
	"time"
)

// CircuitBreaker prevents cascade failures when OpenSearch is unavailable.
type CircuitBreaker struct {
	mu               sync.Mutex
	isOpen           bool
	openUntil        time.Time
	lastErrorPrinted time.Time
	duration         time.Duration
	errorInterval    time.Duration
}

var (
	defaultBreaker     *CircuitBreaker
	defaultBreakerOnce sync.Once
)

// DefaultCircuitBreaker returns the shared circuit breaker instance.
func DefaultCircuitBreaker() *CircuitBreaker {
	defaultBreakerOnce.Do(func() {
		defaultBreaker = NewCircuitBreaker(60*time.Second, 10*time.Second)
	})
	return defaultBreaker
}

// NewCircuitBreaker creates a new circuit breaker with custom durations.
func NewCircuitBreaker(duration, errorInterval time.Duration) *CircuitBreaker {
	return &CircuitBreaker{
		duration:      duration,
		errorInterval: errorInterval,
	}
}

// IsOpen checks if the circuit breaker is currently open.
func (cb *CircuitBreaker) IsOpen() bool {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if !cb.isOpen {
		return false
	}

	// Check if duration has expired
	if time.Now().After(cb.openUntil) {
		return false
	}

	return true
}

// RecordFailure opens the circuit breaker after a failure.
func (cb *CircuitBreaker) RecordFailure(err error) {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	now := time.Now()
	cb.isOpen = true
	cb.openUntil = now.Add(cb.duration)

	// Throttle error printing
	if now.Sub(cb.lastErrorPrinted) > cb.errorInterval {
		fmt.Fprintf(os.Stderr, "[devlogs] Failed to index log, pausing indexing for %.0fs: %v\n",
			cb.duration.Seconds(), err)
		cb.lastErrorPrinted = now
	}
}

// RecordSuccess closes the circuit breaker on successful operation.
func (cb *CircuitBreaker) RecordSuccess() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if cb.isOpen {
		cb.isOpen = false
		fmt.Fprintf(os.Stderr, "[devlogs] Connection restored, resuming indexing\n")
	}
}

File: node/src/circuit-breaker.ts
/**
 * Time-based circuit breaker with auto-recovery.
 *
 * Port of the Go SDK circuit breaker pattern:
 * - Opens on failure, pauses indexing for `duration` (default 60s)
 * - Auto-resets after duration expires (time-based)
 * - Closes immediately on explicit success (RecordSuccess)
 * - Throttles error output to once per `errorInterval` (default 10s)
 */
export class CircuitBreaker {
  private isOpen = false;
  private openUntil = 0;
  private lastErrorPrinted = 0;
  private readonly duration: number;
  private readonly errorInterval: number;
  private readonly stderr: (msg: string) => void;

  /**
   * @param duration - How long to keep circuit open in ms (default: 60000)
   * @param errorInterval - Min interval between error prints in ms (default: 10000)
   * @param stderr - Output function for error messages (default: process.stderr.write)
   */
  constructor(
    duration = 60_000,
    errorInterval = 10_000,
    stderr?: (msg: string) => void
  ) {
    this.duration = duration;
    this.errorInterval = errorInterval;
    this.stderr = stderr ?? ((msg: string) => process.stderr.write(msg + '\n'));
  }

  /**
   * Check if the circuit breaker is open (should skip indexing).
   * Auto-resets if duration has expired.
   */
  shouldSkip(): boolean {
    if (!this.isOpen) return false;
    if (Date.now() >= this.openUntil) {
      this.isOpen = false;
      return false;
    }
    return true;
  }

  /**
   * Record a failure - opens the circuit breaker.
   */
  recordFailure(err: unknown): void {
    const now = Date.now();
    this.isOpen = true;
    this.openUntil = now + this.duration;

    // Throttle error printing
    if (now - this.lastErrorPrinted > this.errorInterval) {
      this.lastErrorPrinted = now;
      const secs = Math.round(this.duration / 1000);
      this.stderr(
        `[devlogs] Failed to index log, pausing indexing for ${secs}s: ${err}`
      );
    }
  }

  /**
   * Record a success - closes the circuit breaker.
   */
  recordSuccess(): void {
    if (this.isOpen) {
      this.isOpen = false;
      this.stderr('[devlogs] Connection restored, resuming indexing');
    }
  }
}
